packages/vat-data/src/vat-data-bindings.js

```diff
@@ -91,19 +93,115 @@ export const partialAssign = (target, source) => {
 // @ts-expect-error TODO statically recognize harden
 harden(partialAssign);
 
+/**
+ * @type {WeakMap<import('./types.js').Baggage, Map<string, (k?:string) => any>>}
+ */
+const collisions = new WeakMap();
+
+/**
+ * HAZARD: `assertUniqueProvide` uses top-level mutable state, but only for
+ * diagnostic purposes.
+ * Aside from being able to provoke it to throw an error, it is not
+ * otherwise usable as a top-level communications channel. The throwing-or-not
+ * of the error is the only way in which this top level state is observable.
+ *
+ * TODO since this still does violate our security assumptions, we should
+ * consider making it development-time only, and turn it off in production.
```

Comment:
A signal of development-time would be useful generally. Do we have a mechanism for that yet? I'm used to using environment variables for that but SES precludes that.

Reply:
Nitpicky terminology note first: For most purposes, "Hardened JS" is the new name for "SES". However, Hardened JS is about the collection of features we expect to propose to standards committees, such as tc39 and tc53. The system we've built we have been calling "the SES-shim". But what does it shim. It includes Hardened JS but also some further abstractions we do not expect to bring to committee, at least not as part of the Hardened JS set of proposals. At the same time, I'm still talking about something narrowing than Endo as a whole. Since the SES-shim shims it, I'll just call it "SES" below.

SES doesn't preclude the use of environment variables for paraterization, and in fact is moving to support it better. The Node mechanism is that environment variables show up in `process.env`. `export FOO=bar` turns into `process.env.FOO === 'bar'`.

SES promotes that into a general convention, supported by https://github.com/endojs/endo/blob/master/packages/ses/src/environment-options.js . It is worth reading the main doc-comment in that module. Right now, this module is only used within SES by `lockdown`, so that lockdown options not provided explicitly in the call to `lockdown` can instead be conditioned on environment variables. For example, my `~/.bashrc` file contains

```sh
export LOCKDOWN_ERROR_TAMING=unsafe
```

so that every call to `lockdown` that does not explicitly provide a setting for `errorTaming` acts like `errorTaming: 'unsafe'` *during development*. The name `'unsafe'` of course was chosen to indicate that this option should never have this setting in production.

The `ses` package currently exports nothing, and @kriskowal would like to keep it that way. But I would like to export `environment-options.js` for use by our code post lockdown. @kriskowal and I have talked about moving it to a distinct package which both ses and others could import. We both find that attractive.

In the meantime, there is plenty of manual direct use of `process.env` in our own code. A bit of grepping shows at least 70 uses that bear further examination. Frankly, it is a bit out of control right now. We need to at least gather an explanation of all of them into one place.

`lockdown` does leave `process.env` intact on the global of the start compartment, which is why the above manual uses are possible. Compartments do not be default pass them through to created compartments. Neither does Swingset pass them through to created vats. Thus they are currently unavailable in vat code running under SwingSet. This is a policy choice. The SwingSet `config` could be enhanced to express other policies about passing environment settings through to the `process.env` of vats running under SwingSet.

Attn @warner @FUDCo @kriskowal 

So with regard to your overall question: On the one hard, we've all worked hard to make our programming environment as deterministic and hermetic as possible, minimizing the observable differences between underlying platforms. OTOH, I've looked for something like a development-time signal myself. We've succeeded so well at the first that I don't currently know of any good way we support the second. This is why I was in this `environment-options.js` code recently, and why @kriskowal and I talked about moving it to a separate package we could all import from. We would still need some kind of SwingSet mechanism to make some difference visible from inside SwingSet vats. We need to be careful how we introduce that.

Separately, we need to systematize and catalog our many manual uses of `process.env` throughout our system and the plethora of invented environment variables they represent.

Reply:
Since the global state does violate our security assumptions, as the comment states, not being able to turn it off for production is a blocker for this PR. I am reluctantly returning this PR to Draft status.

Reply:
Done, according to the plan in the conversation below. Done that is, except it needs a test. I'll track that in the other conversation. Resolving.